File: src/leetcode/ValidSudoku.java
package leetcode;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;

public class ValidSudoku {

    public static void main(String[] args) {
        char[][] board = {
                {'8', '3', '.', '.', '7', '.', '.', '.', '.' },
                {'6', '.', '.', '1', '9', '5', '.', '.', '.' },
                {'.', '9', '8', '.', '.', '.', '.', '6', '.' },
                {'8', '.', '.', '.', '6', '.', '.', '.', '3' },
                {'4', '.', '.', '8', '.', '3', '.', '.', '1' },
                {'7', '.', '.', '.', '2', '.', '.', '.', '6' },
                {'.', '6', '.', '.', '.', '.', '2', '8', '.' },
                {'.', '.', '.', '4', '1', '9', '.', '.', '5' },
                {'.', '.', '.', '.', '8', '.', '.', '7', '9' }};
        System.out.println(isValidSudoku(board));
    }

    public static boolean isValidSudoku(char[][] board) {
        var rows = new HashMap<Integer, Set<Character>>();
        var columns = new HashMap<Integer, Set<Character>>();
        var boxes = new HashMap<Integer, Set<Character>>();

        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                var current = board[r][c];

                if (current == '.')
                    continue;

                int box = (r / 3) * 3 + (c / 3);

                if ((rows.get(r) != null && rows.get(r).contains(current)) ||
                        (columns.get(c) != null && columns.get(c).contains(current)) ||
                        (boxes.get(box) != null && boxes.get(box).contains(current))) {
                    return false;
                }

                rows.computeIfAbsent(r, k -> new HashSet<>()).add(current);
                columns.computeIfAbsent(c, k -> new HashSet<>()).add(current);
                boxes.computeIfAbsent(box, k -> new HashSet<>()).add(current);
            }
        }
        return true;
    }
}

File: src/leetcode/kotlin/ValidSudoku.kt
package leetcode.kotlin

class ValidSudoku {

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            println(
                isValidSudoku(
                    arrayOf(
                        charArrayOf('5', '3', '.', '.', '7', '.', '.', '.', '.'),
                        charArrayOf('6', '.', '.', '1', '9', '5', '.', '.', '.'),
                        charArrayOf('.', '9', '8', '.', '.', '.', '.', '6', '.'),
                        charArrayOf('8', '.', '.', '.', '6', '.', '.', '.', '3'),
                        charArrayOf('4', '.', '.', '8', '.', '3', '.', '.', '1'),
                        charArrayOf('7', '.', '.', '.', '2', '.', '.', '.', '6'),
                        charArrayOf('.', '6', '.', '.', '.', '.', '2', '8', '.'),
                        charArrayOf('.', '.', '.', '4', '1', '9', '.', '.', '5'),
                        charArrayOf('.', '.', '.', '.', '8', '.', '.', '7', '9')
                    )
                )
            )
        }

        fun isValidSudoku(board: Array<CharArray>): Boolean {
            val rows = mutableMapOf<Int, MutableSet<Char>>();
            val columns = mutableMapOf<Int, MutableSet<Char>>();
            val boxes = mutableMapOf<Int, MutableSet<Char>>();

            for (r in 0 until board.size) {
                for (c in 0 until board.size) {
                    val n = board[r][c]

                    if (n == '.') continue

                    val box = (r / 3) * 3 + c / 3

                    if (rows[r]?.contains(n) == true || columns[c]?.contains(n) == true || boxes[box]?.contains(n) == true) {
                        return false
                    }
                    addElement(rows, r, n)
                    addElement(columns, c, n)
                    addElement(boxes, box, n)
                }
            }
            return true
        }

        fun addElement(map: MutableMap<Int, MutableSet<Char>>, key: Int, n: Char) {
            if (map.containsKey(key)) {
                map[key]?.add(n)
            } else {
                map[key] = mutableSetOf(n)
            }
        }
    }
}